packages/react/src/create-strict-api/index.ts

```diff
@@ -0,0 +1,223 @@
+import type { StrictCSSProperties, CSSPseudos } from '../types';
+import { createSetupError } from '../utils/error';
+import { type CompiledStyles, cx, type Internal$XCSSProp } from '../xcss-prop';
+
+type PseudosDeclarations = {
+  [Q in CSSPseudos]?: StrictCSSProperties;
+};
+
+type EnforceSchema<TObject> = {
+  [P in keyof TObject]?: P extends keyof CompiledSchema
+    ? TObject[P] extends Record<string, unknown>
+      ? EnforceSchema<TObject[P]>
+      : TObject[P]
+    : never;
+};
+
+type PickObjects<TObject> = {
+  [P in keyof TObject]: TObject[P] extends Record<string, unknown> ? TObject[P] : never;
+};
+
+interface CompiledAPI<TSchema> {
+  /**
+   * ## CSS
+   *
+   * Creates styles that are statically typed and useable with other Compiled APIs.
+   * For further details [read the documentation](https://compiledcssinjs.com/docs/api-css).
+   *
+   * @example
+   * ```
+   * const redText = css({
+   *   color: 'red',
+   * });
+   *
+   * <div css={redText} />
+   * ```
+   */
+  css(
+    styles: StrictCSSProperties & PseudosDeclarations & EnforceSchema<TSchema>
+  ): StrictCSSProperties;
+  /**
+   * ## CSS Map
+   *
+   * Creates a collection of named styles that are statically typed and useable with other Compiled APIs.
+   * For further details [read the documentation](https://compiledcssinjs.com/docs/api-cssmap).
+   *
+   * @example
+   * ```
+   * const styles = cssMap({
+   *  none: { borderStyle: 'none' },
+   *  solid: { borderStyle: 'solid' },
+   * });
+   *
+   * <div css={styles.solid} />
+   * ```
+   */
+  cssMap<
+    TStyles extends Record<
+      string,
+      StrictCSSProperties & PseudosDeclarations & EnforceSchema<TSchema>
+    >
+  >(
+    styles: TStyles
```

Comment:
@itsdouges I know this will feel redundant, and it is, but I think this will fix the issue (give or take), just copy the `extends` above:
```diff
-    styles: TStyles
+    styles: Record<
+       string,
+       StrictCSSProperties & PseudosDeclarations & EnforceSchema<TSchema>
+     >
```

Currently this is basically just `TStyles` must extend this, which seems right, but technically this **_extends_** `{ color: 'var(--ds-text)' }`:
```
cssMap({
  default: {
    color: 'var(--ds-text)',
    '* > button': { color: 'transparent !important' },
  },
});
```

`extends` is an annoying keyword, I wish there was a `TStyles satisfies …` in generics, but I keep going in circles in TS issues ([like this one](https://github.com/microsoft/TypeScript/issues/51679)) and can't figure out why they don't want to make this change…

Before: [TS Playground(https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBDAnmApnAVHAhgZzgYQGVC4AzKCEOAcgGMcclVqBuAKDZQA9JYFk0RQgAUcKAK4ATCHgC8bOHAA+NAGQAudVlIwUUaguVrNAIyy0A1pIpgDildQ3qTKUtBR2jjzbXEfDDk6+KAC0UCgA5sAQAHaegZqkwFCMIQA2KDC6+gHG6kkpMOnAMf72eRFQWCAgWFAhehQ55d7qtVAWevF5YGnmKAAWEGmSXbmtYhm0MNFx4044qGlpJRENUE3drTB1EZkhulwwW04AbsAoAO4HVTE4wDOxJ1rTwKdlXhpY4jAQScvPEx9GIWZ60AYoSwoSTPUakb5pY7zdRwkrQ2HAHBYIHo5EocBIZ4oGLYjIw5EFRjPNy+HDUiC0kLne44+mMy4PAYlaniZY4WjhYnPIbvZqfdQlMJYGJ7Z4lUbZEAlLC6OUxU5YFbkloaDI6Z4rEEGhma4pG5EQMCPElpZ4QH4hP5SmUfBJgb5ibXisDAabicIhSU+v3hZ69fpDEZ6EI4IaXOY69ThxCrZ7hLCSR0xNKINMoDMhS5QB6u4zhACO4mSuMTxYiAyRiZ2UD2jfF4jE9RKGq1zw70Z7wC9CUHw+MzJLMPYHG4vHgTEExGENj0MxQcjgACV89Js4gADxCAB0y8tq4uOH3jGLMqMMXEIBcUAAfM-pwuCJQfWTCOD8Vg4FkOBCBgYtpiEU9UFgC84FUOBRAkaQcAAEUhPoqmtHBp1naB5wEYDQN9GAIJXaD10Arcd1iHND2II8QOlSQ6kkSDz3IuDj0IU4IlYsicDfDgPwQqQZFQ2h0JVWYNwAb0MABtABFOASgIJcxBEnAAF0AH51AIsDiKXUi1ywtgAF933wgBRGI3CgWgUF-CFan3AAVAB5EwACtIRgZ8KNkxQ5OEZSYjgTpED+OAPO83ydL0kLuF0GJJDwCKovwL9gB-P9akMRRtOizyfOmYLNLgJLiVSyjaGgSQr0ImUABo4HEEEYggeNn3yxQ4EKmy7Icpz-zc4rfLK7ret6vSYpKmAyp6vTSlFdgLME-DMvAbLoV4pAxIk61t1IPRiQcgL5OoAB9S78HcgBZYQAEkABkrOQy7hE3dzhCszdXIATUu5CrPwZ6AEFNzB1zHvcgA5S7NysgAxX6rNh-ArKB9zLth9zXMugB1TdHtczHkMexH8Fc57Acu6hNL00C-FW7CeFw-hUE-LafyQDJL1m3z-KAwK4EU0LwpQSLSCK2LpgZmW5sU8rKpSvBt1qqB6uvVYWraiwOq6nrCs279oRAxA+dG2X5oUzTJumrnTZY4zEH2upJNiI6TpiByWY4Wq7ngAV810Ci3OG2oKqOKq8BN7bJAjrBnwACgASkA-yRcUUg2teWI4HoHA7qwMA4H3Hqptc82+aj5LqvVuqGpvCIWpAgySLPPjYPg9SkLdjCpO7gboCG3KsHDsfXx6lOK96xgLfXGbq-XHrU70rOprgdNdxzUWQpU9Lpar3n13luOeYX-nl5wCb2Cmsy4A3qaAHpn7gAABGAcAaHhfPWJo4AhBCHAS4AwVQoFFLPRQ4QYD+jCvPPmd9epmQ4FNGBcDH4FwYMXUuKCUH+1iIwTBhccFwAfkBYO4D9wb1qmkaAelqAaigMnIBqUDhJVTtQJqPVvAii6OvAuwx6E0CYSwkIbDDhFD4VAThZDzIp1TtOAORCEEoFIRQ7BJdk4bw6qUdeUDBF0KgAw0RrDv6SM4dwzevVeEQFFNQARtDhGMLqGIiRSUQjSNkWZKx1jjAdRgMnfIyRUjgmypIdQASgmUiKGEkYqdOGOKEcYmg4QYRkN8dYxwcB-ImB+L8OISSjEMKBH4agcj77mVTkAA)
After: [TS Playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBDAnmApnAVHAhgZzgYQGVC4AzKCEOAcgGMcclVqBuAKDZQA9JYFk0RQgAUcKAK4ATCHgC8bOHAA+NAGQAudVlIwUUaguVrNAIyy0A1pIpgDildQ3qTKUtBR2jjzbXEfDDk6+KAC0UCgA5sAQAHaegZqkwFCMIQA2KDC6+gHG6kkpMOnAMf72eRFQWCAgWFAhehQ55d7qtVAWevF5YGnmKAAWEGmSXbmtYhm0MNFx4044qGlpJRENUE3drTB1EZkhulwwW04AbsAoAO4HVTE4wDOxJ1rTwKdlXhpY4jAQScvPEx9GIWZ60AYoSwoSTPUakb5pY7zdRwkrQ2HAHBYIHo5EocBIZ4oGLYjIw5EFRjPNy+HDUiC0kLne44+mMy4PAYlaniZY4WjhYnPIbvZqfdQlMJYGJ7Z4lUbZEAlLC6OUxU5YFbkloaDI6Z4rEEGhma4pG5EQMCPElpZ4QH4hP5SmUfBJgb5ibXisDAabicIhSU+v3hZ69fpDEZ6EI4IaXOY69ThxCrZ7hLCSR0xNKINMoDMhS5QB6u4zhACO4mSuMTxYiAyRiZ2UD2jfF4jE9RKGq1zw70Z7wC9CUHw+MzJLMPYHG4vHgTEExGENj0MxQcjgACV89Js4gADxCAB0y8tq4uOH3jGLMqMMXEIBcUAAfM-pwuCJQfWTCOD8Vg4FkOBCBgYtpiEU9UFgC84FUOBRAkaQcAAEUhPoqmtHBp1naB5wEYDQN9GAIJXaD10Arcd1iHND2II8QOlSQ6kkSDz3IuDj0IU4IlYsicDfDgPwQqQZFQ2h0JVWYNwAb0MABtABFOASgIJcxBEnAAF0AH51AIsDiKXUi1ywtgAF933wgBRGI3CgWgUF-CFan3AAVAB5EwACtIRgZ8KNkxQ5OEZSYjgTpED+OAPO83ydL0kLuF0GJJDwCKovwL9gB-P9akMRRtOizyfOmYLNLgJLiVSyjaGgSQr0ImUABo4HEEEYggeNn3yxQ4EKmy7Icpz-zc4rfLK7ret6vSYpKmAyp6vTSlFdgLME-DMvAbLoV4pAxIk61t1IPRiQcgL5OoAB9S78HcgBZYQAEkABkrOQy7hE3dzhCszdXIATUu5CrPwZ6AEFNzB1zHvcgA5S7NysgAxX6rNh-ArKB9zLth9zXMugB1TdHtczHkMexH8Fc57Acu6hNL00C-FW7CeFw-hUE-LafyQDJL1m3z-KAwK4EU0LwpQSLSCK2LpgZmW5sU8rKpSvBt1qqB6uvVYWraiwOq6nrCs279oRAxA+dG2X5oUzTJumrnTZY4zEH2upJNiI6TpiByWY4Wq7ngAV810Ci3OG2oKqOKq8BN7bJAjrBnwACgASkA-yRcUUg2teWI4HoHA7qwMA4H3Hqptc82+aj5LqvVuqGpvCIWpAgySLPPjYPg9SkLdjCpO7gboCG3KsHDsfXx6lOK96xgLfXPSG81pudf0oiO6gkzu+Evu0PdzCh9skfHLHw8spy5yk-txRU70rOprgdNdxzUWQpU9Lpar3n13luOeYXvzau64JrsCmmZOAD8poAHpoFwAAAIwBwA0Hgvl1hNDgCEEIcBLgDBVCgUUs9FDhBgP6MK88+ZgN6mZDgU0SFkMgQXBgxdS40Jof7WIjBGGFxYXACBQFg74P3A-WqaRoB6WoBqKAycsGpQOElVO1Amo9W8CKLo98C7DHETQKRMiQhyMOEUNRUBFF8PMinVO04A5cIoSgXhAjmEl2Tg-DqpR75EM0WIqAEjdGyOQYYxRyjH69VURAUU1ANGiO0ZIuoeiDFJRCMY0xZkgnBOMB1GAyd8jJFSOCbKkh1AZKyZSIoeSRip0UZErR3iaDhBhHw1JwTHBwH8iYH4vw4hVK8RIoEfhqBmPAeZVOQA)

Reply:
Thanks will look soon. I'll say it's supposed to be a feature to block unknown properties (including pseudos) atm 🤔

Reply:
Today it doesn't block them though if your object extends known properties.  It will block if you don't have `color: …` in the above, but once you add it, it won't block.

Before/After playgrounds should be clear…I think I've got this right, I didn't get it all hooked up locally (I don't have this repo setup really).

Reply:
Thanks for the contribution Kylor https://github.com/atlassian-labs/compiled/pull/1582